### ISSUE TITLE
Reproduce and investigate possible race condition in task/undo handling

### ISSUE THREAD

**Issue by dae** [ankitects/anki#2628]
https://forums.ankiweb.net/t/target-undo-op-not-found/33622/18

**Comment by mjuhanne** [comment]
This may be caused when trying to do _col.merge_undo_entries()_ with 30 or more items. 

For example:
```
def add_notes_blocking(col, notes, deck_id, checkpoint):

    print("**** Adding %d notes to checkpoint %s" % (len(notes), checkpoint))
    pos = col.add_custom_undo_entry(checkpoint)

    for note in notes:
        col.add_note(note, deck_id)

    return col.merge_undo_entries(pos)

def add_notes_op(*, notes, deck_id, checkpoint) -> CollectionOp[OpChanges]:
    return CollectionOp(aqt.mw, 
        lambda col, notes=notes,deck_id=deck_id,checkpoint=checkpoint: 
            add_notes_blocking(col, notes, deck_id, checkpoint)
    )

```
If I try to feed this with 30 or more notes, "target undo op not found" error appears.

Then I try to circumvent this by splitting the notes into 29 note batches, each having its own checkpoint (ugly solution, I know):

```
def add_notes(notes, deck_id, checkpoint, success_func):

    maximum_notes_per_checkpoint = 29
    notes_per_checkpoint = [ notes[i:i + maximum_notes_per_checkpoint] 
        for i in range(0, len(notes), maximum_notes_per_checkpoint)
    ]

    checkpoint_number = 0
    for notes_in_checkpoint in notes_per_checkpoint:
        checkpoint_name = checkpoint + '_' + str(checkpoint_number+1)
        print("Creating checkpoint %s" % checkpoint_name)

        op = add_notes_op(notes=notes_in_checkpoint, deck_id=deck_id, checkpoint=checkpoint_name )
        if checkpoint_number == len(notes_per_checkpoint)-1:
            # run success_func only after last batch (checkpoint) is processed
            op.success(success_func)
        op.run_in_background()
        checkpoint_number += 1
```

This may work for 35-40 notes but when adding 300 notes with concurrent 5 background processes I get a bunch of 'target undo op not found' messages again.

Is there a way to do this elegantly and without errors?

**Comment by dae** [comment]
Your code should work once we fix the issue I reported here. I suspect the problem is the executor in taskman.py will spawn multiple threads. We could pass it a max_workers=1, but then when doing non-collection stuff like fetching files from the internet, the requests would end up serialized, which is undesirable.  A quick-and-dirty fix for this might be to have CollectionOp's wrapped_op() take a lock on a global mutex, and drop it after the inner command is run.

**Comment by abdnh** [comment]
>  A quick-and-dirty fix for this might be to have CollectionOp's wrapped_op() take a lock on a global mutex, and drop it after the inner command is run.

That works with @mjuhanne's example (I can't say that for @L-M-Sherlock's situation since I couldn't trigger the error with his [reproduction steps](https://forums.ankiweb.net/t/target-undo-op-not-found/33622/20)).


Any other reason not to go with the global mutex solution if CollectionOps are not meant to interleave?

**Comment by dae** [comment]
The alternative would be to run collection ops on a second pool that has max_workers=1. That would avoid any accidental forgot-to-unlock situations, but they're unlikely if we're only locking in one place, and would require updates to taskman, so using a mutex works for me too.

**Comment by L-M-Sherlock** [comment]
> I couldn't trigger the error with his [reproduction steps](https://forums.ankiweb.net/t/target-undo-op-not-found/33622/20)

I find that the error occurred occasionally.

**Comment by dae** [comment]
One possible argument for a separate task pool - there are some operations like a colpkg import that we'd also like serialized with collection ops, but they're not collection ops themselves.

**Comment by dae** [comment]
I'll try tackle this sometime in the next few days, assuming you haven't already started on it @abdnh.

**Comment by abdnh** [comment]
Will leave that to you then (I remembered I've been planning to send #2659 for a while)

**Comment by dae** [comment]
@L-M-Sherlock @mjuhanne, can you confirm this fixes your issue when you run the latest commit on main

**Comment by L-M-Sherlock** [comment]
<img width="667" alt="image" src="https://github.com/ankitects/anki/assets/32575846/e72a54cc-a6b5-4085-9b6b-9f75c4c78e51">

It still exists. I plan to write an minimal add-on to reproduce it.

**Comment by L-M-Sherlock** [comment]
Have a try:

[reproduce.zip](https://github.com/ankitects/anki/files/12616305/reproduce.zip)

**Comment by dae** [comment]
Thanks, that was helpful. I think this should now be fixed.

**Comment by luc-languagetools** [comment]
HyperTTS is experiencing the "target undo op not found" issue occasionally on Anki 23.12.1.  Here's the code: 
https://github.com/Language-Tools/anki-hyper-tts/blob/main/hypertts.py#L51
`undo_id = self.anki_utils.undo_start()` simply does `aqt.mw.col.add_custom_undo_entry('hypertts')`, and `self.anki_utils.undo_end(undo_id)` does:
```
aqt.mw.col.merge_undo_entries(undo_id)
aqt.mw.update_undo_actions()
aqt.mw.autosave()
```

It's a bit tricky to reproduce, usually involves batch generating audio on a large number of notes (hundreds), but I'm not able to reproduce it consistently. It's been a challenge to make progress on this issue.

**Comment by dae** [comment]
The operation should be called inside a CollectionOp() so it doesn't run in parallel with other operations.

**Comment by luc-languagetools** [comment]
I'm still confused about how I can run updates on a bunch of notes and still group them into a single undo entry. Will this not require the following steps  (currently all run in the background) ?
1. start a new undo entry
```
undo_id = aqt.mw.col.add_custom_undo_entry('hypertts')
```
2. update notes in a loop
```
for note_id in note_id_list:
  # do some processing, update note (for example add audio tag)
  aqt.mw.col.update_note(note)
```
3. merge undo entries
```
aqt.mw.col.merge_undo_entries(undo_id)
aqt.mw.update_undo_actions()
aqt.mw.autosave()
```

This is my basic workflow and it currently occasionally gives me the result `target undo op not found`, particularly with large number of notes like > 300. Your suggestion is to wrap part 2 in a CollectionOp, which i'm a bit confused about, because i'm already in a background task, it's not obvious to me how to execute the CollectionOp when currently running a background task.

It might be safer to collect changes notes objects and then do `aqt.operations.note.update_notes` at step 2, does that reduce the likelyhood of getting `target undo op not found` ? Given step 1, 2, 3 are not atomic, is there still a potential race condition where the undo_id could get lost ?

**Comment by dae** [comment]
The entire operation - steps 1 to 3 - should be run in a collection op, with the value returned by merge_undo_entries being returned. That will prevent any other well-behaved tasks from executing in parallel, and will correctly update the UI when the operation completes. The other two lines in 3 are not required.

**Comment by luc-languagetools** [comment]
OK that seems to do the trick, thank you @dae . I had not clearly understood that CollectionOps were meant to serialize collection access. It all makes sense now and I think I know how i'll be integrating this loop into a CollectionOp.

**Comment by dae** [comment]
IIRC an extra progress bar won't be shown if there's one already active. If not, you'd need to use QueryOp instead.

**Comment by luc-languagetools** [comment]
QueryOp works perfectly. HyperTTS has its own progress bar embedded into the UI. Many thanks for your help !
<img width="901" alt="62134d389325c421e5189a80_add_audio_collection_apply_to_notes" src="https://github.com/ankitects/anki/assets/4581915/99d8297d-3733-4292-b916-d5b25589fcc9">
